src/inmanta/execute/dataflow.py

```diff
@@ -60,25 +213,63 @@ def own_instance_node_for_responsible(
         return self._own_instances[responsible]
 
     def _add_global_instance_node(self, node: "InstanceNodeReference") -> None:
-        # TODO: implement
-        pass
+        """
+            Registers an instance.
+        """
+        entity: Optional["Entity"] = node.top_node().entity
+        if entity is not None:
+            self._entity_data(entity).add_instance(node)
 
-    def add_bidirectional_attribute(self, entity: "Entity", this: str, other: str) -> None:
-        # TODO: implement
-        pass
+    def register_bidirectional_attribute(self, entity: "Entity", this: str, other: str) -> None:
+        """
+            Registers a pair of bidirectional attributes.
+        """
+        self._entity_data(entity).register_bidirectional_attribute(this, other)
 
     def add_index_match(self, instances: Iterable["InstanceNodeReference"]) -> None:
-        # TODO: implement
-        pass
+        """
+            Registers an index match between a set of instance nodes. Informs the nodes of the match.
+        """
+        iterator: Iterator[InstanceNodeReference] = iter(instances)
+        try:
+            first: InstanceNode = next(iter(iterator)).node()
+            for instance in iterator:
+                instance.node().index_match(first)
+        except StopIteration:
+            pass
+
 
-    # TODO: other methods
+class EntityData:
+    """
+        Data about an entity. Contains an entity's bidirectional attributes and a list of all instances.
+        Enforces bidirectionality for all instances.
+    """
+
+    __slots__ = ("entity", "bidirectional_attributes", "instances")
+
+    def __init__(self, entity: "Entity") -> None:
+        self.entity: "Entity" = entity
+        self.bidirectional_attributes: Dict[str, str] = {}
+        self.instances: List[InstanceNodeReference] = []
+
+    def register_bidirectional_attribute(self, this: str, other: str) -> None:
+        self.bidirectional_attributes[this] = other
+        for instance in self.instances:
```

Comment:
This has already been fixed as I believe much as possible. Instances no longer store their bidirectional attributes, but the InstanceNodes that already exist at this point still need to be informed of them in case they've already got assignments for those attributes. It might be a good idea to rename the method on the InstanceNodr side though so it's clear it doesn't store them, just processes them.

Reply:
The registration could be done in the evaluate of DefineRelation instead of the emit, so that all relations are know prior to execution

Reply:
I think that actually might be the case already, but that would break this module's independency (mostly) of compiler scheduling. Currently any statement can be adde at any time, which makes for a nice decoupling of the logic. I'm not necessarily saying we shouldn't do what you suggest, but we do need to be aware of the consequences.